### ISSUE TITLE
Gradient of gradient explodes(nan) when training WGAN-GP on Mnist

### ISSUE THREAD

**Issue by MeJn** [pytorch/pytorch#2534]
I was using WGAN-GP training to generate handwritten character images on Mnist dataset. Thanks for [caogang's code](https://github.com/caogang/wgan-gp), I modified his network structure, the DCGAN was applied to my WGAN-GP's structure.
During the training process, a sudden explosion (**nan**) of the gradients occurred, and the location of the explosion was after the backward propagation using the gradient penalty loss.

Before the explosion, the gradients and the parameters was always normal and there was no tendency to gradually increase, so I thought it was the problem of calculating the gradient of the gradient.

**Comment by LukasMosser** [comment]
@MeJn I'm experiencing the same thing happening. I've tried different hyperparameters BN vs. no BN etc. but at some point the gradient always explodes to nan. Maybe there's some reset that needs to be done?

**Comment by MeJn** [comment]
@LukasMosser I'v also tried the same thing as you do, and the result was same as yours too. I think it may be related to the [2421 issue](https://github.com/pytorch/pytorch/issues/2421).

**Comment by dannysdeng** [comment]
Similar problem found. I added eps before calculating the norm() in WGAN-GP and it seems to mitigate at least the NaN problem.

    x = Variable(torch.zeros(1), requires_grad=True)
    x = x + 1e-16
    x.norm().backward()
    print x.grad

**Comment by MeJn** [comment]
@dannysdeng I'v tried this method, but the same problem NaN occurred, how did you add the eps? would you mind sharing the code for calculating Gradient Penalty to me? Thank you very much!

**Comment by dannysdeng** [comment]
@MeJn Hi, my code looks like this. It seems to relieve the NaN problem for me. (If "x" is zero, then the gradient of "x.norm()" will be equal to 1)

    alpha = torch.rand(batch_size, 1).expand(X.size())
    x_hat = Variable(alpha * X.data + (1 - alpha) * (X.data + 0.5 * X.data.std() * torch.rand(X.size())), 
    requires_grad=True)
    pred_hat = netD(x_hat)
    gradients = grad(outputs=pred_hat, inputs=x_hat, grad_outputs=torch.ones(pred_hat.size()),
                        create_graph=True, retain_graph=True, only_inputs=True)[0]
    # gradient_penalty = lambda_ * ((gradients.norm(2, dim=1) - 1) ** 2).mean()
    gradients_debug = gradients + 1e-16
    gradient_penalty = lambda_ * ((gradients_debug.norm(2, dim=1) - 1) ** 2).mean()
    gradient_penalty.backward()

**Comment by PiotrDabkowski** [comment]
@soumith  I have the same problem. I thought that reducing the contribution of the gradient penalty may solve the problem, but no! Even setting the contribution to exactly 0 causes NaN:

```
gr = torch.autograd.grad(mid_scores, mids, create_graph=True)[0] + 0.0001
gradient_penalty = ((gr.norm(dim=0) - 1.)**2).mean()
loss = fake_scores - real_scores + self.drift_penalty_factor*(real_scores**2)  + 0.*gradient_penalty
```

Note that the gradient penalty is multiplied by 0 so should not have any impact on the optimisation. However, even when multiplied by 0 it causes some kind of huge numerical instability and the only solution is to not include it in the loss term at all.

**Comment by apaszke** [comment]
@PiotrDabkowski 0 * NaN = NaN, inf * 0 = NaN. This multiplication won't help with any gradients that aren't finite.

**Comment by PiotrDabkowski** [comment]
@apaszke Thanks, yes, I just wanted to check whether the loss term causes the model to be unstable by driving the parameters toward inf. But as you said the gradient of gradient_penalty is either NaN or inf even with finite values of parameters. The gradient_penalty itself is not NaN (only its gradient).

The solution proposed by @dannysdeng does not work as the gradient of:

    gradient_penalty = torch.sum(gr)

is also NaN, so the problem is not with norm. 

I noticed that swapping the  ReLU activation with LeakyReLU solves the problem, but I am not entirely sure why. I will try to produce a minimal example where the gradient calculation is unstable.

**Comment by yxchng** [comment]
@PiotrDabkowski Is there any tradeoff with using eps=1e-8 or eps=1e-5? I am able to avoid NaN with 1e-5 but I am still getting NaN with 1e-8. How should I check if it is fine with 1e-5?

**Comment by zhouhanc** [comment]
I'm having the same problem. Does anyone have a good workaround? 

Thanks!

**Comment by soumith** [comment]
according to @PiotrDabkowski 's comment above, he found the reason and a workaround. Hence, closing.
